[InstCombine] Fold vector.reduce.op(vector.reverse(X)) -> vector.reduce.op(X)

david-arm opened the pull request:
For all of the following reductions:

vector.reduce.or
vector.reduce.and
vector.reduce.xor
vector.reduce.add
vector.reduce.mul
vector.reduce.umin
vector.reduce.umax
vector.reduce.smin
vector.reduce.smax
vector.reduce.fmin
vector.reduce.fmax

if the input operand is the result of a vector.reverse then we can perform a reduction on the vector.reverse input instead since the answer is the same. If the reassociation is permitted we can also do the same folds for these:

vector.reduce.fadd
vector.reduce.fmul

david-arm committed:
[InstCombine] Fold vector.reduce.op(vector.reverse(X)) -> vector.reduce.op(X)

For all of the following reductions:

vector.reduce.or
vector.reduce.and
vector.reduce.xor
vector.reduce.add
vector.reduce.mul
vector.reduce.umin
vector.reduce.umax
vector.reduce.smin
vector.reduce.smax
vector.reduce.fmin
vector.reduce.fmax

if the input operand is the result of a vector.reverse then we
can perform a reduction on the vector.reverse input instead since
the answer is the same. If the reassociation is permitted we can
also do the same folds for these:

vector.reduce.fadd
vector.reduce.fmul

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -3394,8 +3436,31 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+      }
+      return replaceInstUsesWith(CI, Res);
+    }
+
```
I think a general easier way to implement this would be create a helper in `InstuctionCombiner`. The code for all them can simply be `replaceOperand` of the reduce with the pre-reversed vec. That will also make it easier to expand (for example any shuffle that preserves all elements).

davemgreen replied:
The existing shuffle code is down below, and I agree it would be nice for them to be treated similarly.

david-arm replied:
OK that makes sense. I'll have a go!

david-arm replied:
There is a problem with this approach. In the Intrinsic::vector_reduce_or case if I use the same approach as we do for fadd, etc. then what happens is on the first iteration we hit this code:

```
      replaceUse(II->getOperandUse(ArgIdx), V);
      return nullptr;
```

and then on the second iteration we still hit the existing code in Intrinsic::vector_reduce_or:

```
     if (!IsReverse && match(Arg, m_ZExtOrSExtOrSelf(m_Value(Vect)))) {
       ...
          return replaceInstUsesWith(CI, Res);
```

because on the 2nd iteration the reduce_or input is no longer a reverse intrinsic. InstCombine quite rightly gets unhappy with this error:

LLVM ERROR: Instruction Combining did not reach a fixpoint after 1 iterations

Essentially it looks like the only way I can stop the iterations is to return `replaceInstUsesWith` when we first spot the pattern (reduce.or(vector.reverse())) unless you have other suggestions that may help?

david-arm replied:
If it helps I can post the broken patch on this PR so you can see the code and find out if I've done something wrong.

david-arm replied:
Perhaps the solution is to return the original instruction instead of `nullptr` as this might break the loop?

david-arm committed:
Add simplifyReductionOfShuffle helper

* Let all of the vector.reduce.op variants now call a new helper
called simplifyReductionOfShuffle, which deals with both
vector.reverse and shufflevector operations.

paulwalker-arm commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1435,6 +1435,51 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+Instruction *InstCombinerImpl::simplifyReductionOfShuffle(IntrinsicInst *II) {
+  Intrinsic::ID IID = II->getIntrinsicID();
+  bool CanBeReassociated = (IID != Intrinsic::vector_reduce_fadd &&
+                            IID != Intrinsic::vector_reduce_fmul) ||
+                           II->hasAllowReassoc();
+
+  if (!CanBeReassociated)
+    return nullptr;
+
+  const unsigned ArgIdx = (IID == Intrinsic::vector_reduce_fadd ||
+                           IID == Intrinsic::vector_reduce_fmul)
+                              ? 1
+                              : 0;
+  Value *Arg = II->getArgOperand(ArgIdx);
+  Value *V;
+
```
Just a thought but perhaps this will be cleaner is you created something like `simplifyReductionOperand(Value* Op, bool CanReorderLanes)`?  So you'd end up with either:
```
if (auto *NewOp = simplifyReductionOperand(II->getArgOperand(0), true) {
  replaceUse(II->getOperandUse(0), NewOp);
  return nullptr;
}

if (auto *NewOp = simplifyReductionOperand(II->getArgOperand(1), II->hasAllowReassoc()) {
  replaceUse(II->getOperandUse(1), NewOp);
  return nullptr;
}
```

paulwalker-arm replied:
That's not quite what I meant.  I was more thinking `simplifyReductionOperand` would take the reduction's operand rather than the reduction itself.  This way the reduction specific parts remain at the call site.

david-arm replied:
Done

david-arm commented:
Rebase

david-arm committed:
Rename simplifyReductionOfShuffle -> simplifyReductionOperand

... and pass in a flag indicating whether we can reorder lanes.

david-arm committed:
Address review comment

* Pass reduction operand to simplifyReductionOperand instead and
return the simplified operand.

paulwalker-arm commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1435,6 +1435,34 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+Value *InstCombinerImpl::simplifyReductionOperand(Value *Arg,
+                                                  bool CanReorderLanes) {
```
I'm happy either way but I suppose at this moment in time this could just be a static function.

david-arm replied:
Done

david-arm committed:
Make simplifyReductionOperand static

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -3364,6 +3421,12 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
       Value *Arg = II->getArgOperand(0);
       Value *Vect;
+
+      if (Value *NewOp = simplifyReductionOperand(Arg, true)) {
```
Nit: Here and above, style is `/*CanReorderLanes=*/true`

david-arm replied:
Done!

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1435,6 +1435,33 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
+
+  // Can remove shuffle iff just shuffled elements, no repeats, undefs, or
+  // other changes.
+  return UsedIndices.all() ? V : nullptr;
```
Think the shuffle check could be simplified with `getShuffleDemandedElts`.

Then the check would just be:
```
if(DemandedLHS.isAllOnes() && DemandedRHS.isZero()) {
  return LHS;
}
if(DemandedRHS.isAllOnes() && DemandedLHS.isZero()) {
  return RHS;
}
return nullptr;
```

david-arm replied:
Hmm, this is not a NFC change. I am not quite sure how to use `getShuffleDemandedElts`, but I'm pretty sure your example above now permits use of a RHS whereas previously it didn't (see `!match(Arg, m_Shuffle(m_Value(V), m_Undef(), m_Mask(Mask)))`). I would prefer to limit this to refactoring only as it's not relevant to this patch. I can keep the `if(DemandedLHS.isAllOnes() ...` check.

david-arm replied:
I had a look at using `getShuffleDemandedElts` and the code didn't immediately look much better as I still need to keep most of the existing checks, i.e. check for `isSingleSource` and so on. For now, I'd prefer to leave the code as it is, but I'm happy to revisit this in a follow-on patch. Just out of curiosity, have you seen any examples where we pass through the 2nd operand of the shufflevector? i.e. `reduce_or(shufflevector(<4 x i32> %a, <4 x i32> %b, <4 x i32> <7, 5, 6, 4>))`. I'd expect this to be canonicalised to `reduce_or(shufflevector(<4 x i32> %a, <4 x i32> undef, <4 x i32> <3, 1, 2, 0>))`

david-arm committed:
Add comment to simplifyReductionOperand argument
